### PR TITLE
expected responsibilities for each role backend

### PR DESCRIPTION
**Pull request by ciarasugrue**

### CODE REVIEW

File: src/main/java/org/kainos/ea/api/ResponsibilityService.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public List<Responsibility> getAllResponsibilities() throws FailedToGetResponsibilitiesException {
+        try {
+            return responsibilityDao.getAllResponsibilities(databaseConnector.getConnection());
+        } catch (SQLException e) {
+            System.err.println(e.getMessage());
+            throw new FailedToGetResponsibilitiesException();
+        }
+    }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/api/ResponsibilityService.java]: not sure this method is needed?

File: src/main/java/org/kainos/ea/db/ResponsibilityDao.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                Responsibility responsibility = new Responsibility(
+                        results.getInt("responsibility_id"),
+                        results.getString("responsibility_desc")
+                );
+                allResponsibilities.add(responsibility);
+            }
+
+            return allResponsibilities;
+        }
+    }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/db/ResponsibilityDao.java]: not sure this method is needed?

File: src/main/java/org/kainos/ea/resources/ResponsibilityController.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    @Path("/responsibilities")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getAllResponsibilities() {
+        try {
+            List<Responsibility> responsibilities = responsibilityService.getAllResponsibilities();
+            return Response.ok(responsibilities).build();
+        } catch (FailedToGetResponsibilitiesException e) {
+            return Response.serverError().entity(e.getMessage()).build();
+        }
+    }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/resources/ResponsibilityController.java]: not sure this method is needed?

File: src/test/java/org/kainos/ea/controller/ResponsibilityControllerUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import javax.ws.rs.core.Response;
+import java.util.Arrays;
+import java.util.List;
+
+public class ResponsibilityControllerUnitTest {
+
+    ResponsibilityService responsibilityService = Mockito.mock(ResponsibilityService.class);
+
+    @Test
+    void getAllResponsibilities_shouldReturn200StatusCodeAndListOfResponsibilities_whenServiceReturnsListOfResponsibilities() throws FailedToGetResponsibilitiesException {
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/controller/ResponsibilityControllerUnitTest.java]: add unit tests for getResponsibilitiesForJob method as well

File: src/main/java/org/kainos/ea/api/ResponsibilityService.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    public List<Responsibility> getResponsibilitiesForJob(int jobId) throws FailedToGetResponsibilitiesException {
+        try {
+            return responsibilityDao.getResponsibilitiesForJob(databaseConnector.getConnection(), jobId);
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/api/ResponsibilityService.java:31]: can you validate that the jobId exists in db before calling dao

File: src/main/java/org/kainos/ea/resources/ResponsibilityController.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    @GET
+    @Path("/jobs/{jobId}/responsibilities")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getResponsibilitiesForJob(@PathParam("jobId") int jobId) {
+        try {
+            List<Responsibility> responsibilities = responsibilityService.getResponsibilitiesForJob(jobId);
+            return Response.ok(responsibilities).build();
+        } catch (FailedToGetResponsibilitiesException e) {
+            return Response.serverError().entity(e.getMessage()).build();
+        }
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/resources/ResponsibilityController.java:59]: should also have a 404 being returned for job not found

File: src/test/java/org/kainos/ea/service/ResponsibilityServiceUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+@ExtendWith(MockitoExtension.class)
+public class ResponsibilityServiceUnitTest {
+
+    ResponsibilityDao responsibilityDao = Mockito.mock(ResponsibilityDao.class);
+    DatabaseConnector databaseConnector = Mockito.mock(DatabaseConnector.class);
+
+    ResponsibilityService responsibilityService = new ResponsibilityService(responsibilityDao, databaseConnector);
+    Connection connection;
+
+    @Test
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/service/ResponsibilityServiceUnitTest.java:32]: add tests for getResponsibilitiesForJob method

**Review by shaunganley — CHANGES_REQUESTED** [review]
you will also need some integration tests for new endpoint

File: src/main/java/org/kainos/ea/api/ResponsibilityService.java (hunk 1 above)
Comment (ciarasugrue) [reply src/main/java/org/kainos/ea/api/ResponsibilityService.java]: Removed

File: src/main/java/org/kainos/ea/db/ResponsibilityDao.java (hunk 2 above)
Comment (ciarasugrue) [reply src/main/java/org/kainos/ea/db/ResponsibilityDao.java]: Removed

File: src/main/java/org/kainos/ea/resources/ResponsibilityController.java (hunk 3 above)
Comment (ciarasugrue) [reply src/main/java/org/kainos/ea/resources/ResponsibilityController.java]: Removed

File: src/main/java/org/kainos/ea/api/ResponsibilityService.java (hunk 5 above)
Comment (ciarasugrue) [reply src/main/java/org/kainos/ea/api/ResponsibilityService.java:31]: Method to check if job exists added in the Dao, and invoked in the getResponsibilitiesForJob method in ResponsibilityService class.

File: src/main/java/org/kainos/ea/resources/ResponsibilityController.java (hunk 6 above)
Comment (ciarasugrue) [reply src/main/java/org/kainos/ea/resources/ResponsibilityController.java:59]: Added

File: src/test/java/org/kainos/ea/controller/ResponsibilityControllerUnitTest.java (hunk 4 above)
Comment (ciarasugrue) [reply src/test/java/org/kainos/ea/controller/ResponsibilityControllerUnitTest.java]: unit tests added

File: src/test/java/org/kainos/ea/service/ResponsibilityServiceUnitTest.java (hunk 7 above)
Comment (ciarasugrue) [reply src/test/java/org/kainos/ea/service/ResponsibilityServiceUnitTest.java:32]: getAllResponsibilities test have been removed, as the method has been removed from the Dao. Tests added for getResponsibilitiesForJob method

File: src/test/java/org/kainos/ea/integration/ResponsibilityControllerIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+//import io.dropwizard.configuration.ResourceConfigurationSourceProvider;
+//import org.junit.jupiter.api.Assertions;
+//import org.junit.jupiter.api.Test;
+//import org.kainos.ea.FloorIsJavaApplication;
+//import org.kainos.ea.FloorIsJavaConfiguration;
+//import io.dropwizard.testing.junit5.DropwizardAppExtension;
+//import io.dropwizard.testing.junit5.DropwizardExtensionsSupport;
+//import org.junit.jupiter.api.extension.ExtendWith;
+//import org.kainos.ea.cli.Responsibility;
+//
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/integration/ResponsibilityControllerIntegrationTest.java]: remove commented code

File: src/test/java/org/kainos/ea/service/ResponsibilityServiceUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    public void getResponsibilitiesForJob_shouldThrowFailedToGetResponsibilitiesException_whenDaoThrowsSQLException() throws SQLException {
+        // sample job id
+        int jobId = 1;
+
+        // simulation of job not found in the database
+        Mockito.when(responsibilityDao.doesJobExist(connection, jobId)).thenReturn(false);
+
+        assertThrows(FailedToGetResponsibilitiesException.class,
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/service/ResponsibilityServiceUnitTest.java]: this should throw a not found exception instead which should return a 404 from controller

File: src/test/java/org/kainos/ea/integration/ResponsibilityControllerIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+@ExtendWith(DropwizardExtensionsSupport.class)
+public class ResponsibilityControllerIntegrationTest {
+
+    static final DropwizardAppExtension<FloorIsJavaConfiguration> APP = new DropwizardAppExtension<>(
+            FloorIsJavaApplication.class, null,
+            new ResourceConfigurationSourceProvider()
+    );
+
+    @Test
+    void getResponsibilitiesForJob_ShouldReturnListOfResponsibilitiesForJob() {
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/integration/ResponsibilityControllerIntegrationTest.java:41]: add an integration test for a job id that doesn't exist and expect it to return a 404

File: src/main/java/org/kainos/ea/resources/ResponsibilityController.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    @Path("/jobs/{jobId}/responsibilities")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getResponsibilitiesForJob(@PathParam("jobId") int jobId) {
+        try {
+            List<Responsibility> responsibilities = responsibilityService.getResponsibilitiesForJob(jobId);
+            if (responsibilities.isEmpty()) {
+                return Response.status(Response.Status.NOT_FOUND).entity("Job with ID " + jobId + " not found").build();
+            }
+            return Response.ok(responsibilities).build();
+        } catch (FailedToGetResponsibilitiesException e) {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/resources/ResponsibilityController.java:54]: return a 404 if job id not found

File: src/test/java/org/kainos/ea/controller/ResponsibilityControllerUnitTest.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        ResponsibilityController responsibilityController = new ResponsibilityController(responsibilityService);
+
+        Response response = responsibilityController.getResponsibilitiesForJob(jobId);
+
+        Assertions.assertEquals(response.getStatus(), 200);
+        Assertions.assertEquals(response.getEntity(), expectedResponseBody);
+    }
+
+    @Test
+    void getResponsibilitiesForJob_shouldReturn500StatusCodeAndErrorMessage_whenServiceThrowsFailedToGetResponsibilitiesException() throws FailedToGetResponsibilitiesException {
```
Comment (shaunganley) [inline src/test/java/org/kainos/ea/controller/ResponsibilityControllerUnitTest.java]: add a test to return a 404 if job not found exception is thrown

File: src/main/java/org/kainos/ea/api/ResponsibilityService.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    public ResponsibilityService(ResponsibilityDao responsibilityDao, DatabaseConnector databaseConnector) {
+        this.responsibilityDao = responsibilityDao;
+        this.databaseConnector = databaseConnector;
+    }
+
+    public List<Responsibility> getResponsibilitiesForJob(int jobId) throws FailedToGetResponsibilitiesException {
+        try {
+            // Check if jobId exists in the database before calling the dao
+            if (!responsibilityDao.doesJobExist(databaseConnector.getConnection(), jobId)) {
+                throw new FailedToGetResponsibilitiesException();
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/api/ResponsibilityService.java]: this should be a job not found exception

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: src/test/java/org/kainos/ea/integration/ResponsibilityControllerIntegrationTest.java (hunk 8 above)
Comment (ciarasugrue) [reply src/test/java/org/kainos/ea/integration/ResponsibilityControllerIntegrationTest.java]: removed

File: src/test/java/org/kainos/ea/service/ResponsibilityServiceUnitTest.java (hunk 9 above)
Comment (ciarasugrue) [reply src/test/java/org/kainos/ea/service/ResponsibilityServiceUnitTest.java]: Updated

File: src/main/java/org/kainos/ea/resources/ResponsibilityController.java (hunk 11 above)
Comment (ciarasugrue) [reply src/main/java/org/kainos/ea/resources/ResponsibilityController.java:54]: Done.

File: src/main/java/org/kainos/ea/api/ResponsibilityService.java (hunk 13 above)
Comment (ciarasugrue) [reply src/main/java/org/kainos/ea/api/ResponsibilityService.java]: JobNotFoundException class created and invoked here

**Review by eoinmcbennett — APPROVED** [review]

File: src/main/java/org/kainos/ea/api/ResponsibilityService.java [hunk 14]
```diff
@@ -32,4 +34,16 @@ public List<Responsibility> getResponsibilitiesForJob(int jobId) throws FailedTo
             throw new FailedToGetResponsibilitiesException();
         }
     }
+    // New method to check if a job exists
+    public boolean doesJobExist(int jobId) throws SQLException {
+        Connection connection = null;
+        try {
+            connection = databaseConnector.getConnection();
+            return responsibilityDao.doesJobExist(connection, jobId);
+        } finally {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/api/ResponsibilityService.java]: you don't need to do this, we can re-use the connection

File: src/main/java/org/kainos/ea/resources/ResponsibilityController.java [hunk 15]
```diff
@@ -20,29 +24,35 @@
...
         this.responsibilityService = responsibilityService;
     }
 
     @GET
     @Path("/jobs/{jobId}/responsibilities")
     @Produces(MediaType.APPLICATION_JSON)
-    public Response getResponsibilitiesForJob(@PathParam("jobId") int jobId) {
+    public Response getResponsibilitiesForJob(@PathParam("jobId") int jobId) throws SQLException {
         try {
+            if (!responsibilityService.doesJobExist(jobId)) {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/resources/ResponsibilityController.java]: it looks like you're already doing this check in your service method? can you remove it from here?

File: src/main/java/org/kainos/ea/resources/ResponsibilityController.java [hunk 16]
```diff
@@ -20,29 +24,35 @@
...
+                //otherwise find responsibilities
             List<Responsibility> responsibilities = responsibilityService.getResponsibilitiesForJob(jobId);
+                //return 404 if no responsibilities found
             if (responsibilities.isEmpty()) {
                 return Response.status(Response.Status.NOT_FOUND).entity("Job with ID " + jobId + " not found").build();
             }
+                //otherwise return status 200 okay
             return Response.ok(responsibilities).build();
-        } catch (FailedToGetResponsibilitiesException e) {
+        } catch (FailedToGetResponsibilitiesException | NotFoundException e) {
```
Comment (shaunganley) [inline src/main/java/org/kainos/ea/resources/ResponsibilityController.java]: NotFoundException should return 404

**Review by shaunganley — CHANGES_REQUESTED** [review]

**Review by shaunganley — APPROVED** [review]

File: src/main/java/org/kainos/ea/resources/ResponsibilityController.java (hunk 16 above)
Comment (ciarasugrue) [reply src/main/java/org/kainos/ea/resources/ResponsibilityController.java]: Done.

File: src/main/java/org/kainos/ea/resources/ResponsibilityController.java (hunk 15 above)
Comment (ciarasugrue) [reply src/main/java/org/kainos/ea/resources/ResponsibilityController.java]: Removed.

File: src/main/java/org/kainos/ea/api/ResponsibilityService.java (hunk 14 above)
Comment (ciarasugrue) [reply src/main/java/org/kainos/ea/api/ResponsibilityService.java]: Done.

**Pull request merged**